beacon-chain/sync/data_columns_reconstruct.go

```diff
@@ -220,3 +220,198 @@ func (s *Service) broadcastMissingDataColumnSidecars(
 
 	return nil
 }
+
+// processDataColumnSidecarsFromExecutionFromBlock retrieves (if available) data column sidecars data from the execution client,
+// builds corresponding sidecars, save them to the storage, and broadcasts them over P2P if necessary.
+func (s *Service) processDataColumnSidecarsFromExecutionFromBlock(ctx context.Context, roSignedBlock interfaces.ReadOnlySignedBeaconBlock) error {
+	secondsPerHalfSlot := time.Duration(params.BeaconConfig().SecondsPerSlot/2) * time.Second
+	block := roSignedBlock.Block()
+
+	// Exit early if there is no commitments in the block.
+	commitments, err := block.Body().BlobKzgCommitments()
+	if err != nil {
+		return errors.Wrap(err, "blob kzg commitments block")
+	}
+	if len(commitments) == 0 {
+		return nil
+	}
+
+	slot := block.Slot()
+	proposerIndex := block.ProposerIndex()
+
+	// Compute indices of data columns we should sample.
+	indices, err := s.columnIndicesToSample()
+	if err != nil {
+		return errors.Wrap(err, "column indices to sample")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, secondsPerHalfSlot)
+	defer cancel()
+
+	for {
+		// Return early if all required data columns are already seen.
+		if s.haveAllSidecarsBeenSeen(slot, proposerIndex, indices) {
+			return nil
+		}
+
+		// Try to reconstruct data column sidecars from the execution client.
+		sidecars, err := s.cfg.executionReconstructor.ConstructDataColumnSidecarsFromBlock(ctx, roSignedBlock)
+		if err != nil {
+			return errors.Wrap(err, "reconstruct data column sidecars")
+		}
+
+		// Broadcast and save data column sidecars to custody but not yet received.
+		// Wait for a short delay and return if `sidecars` is empty.
+		if err := s.broadcastAndReceiveUnseenSidecars(ctx, slot, proposerIndex, sidecars, indices, "beaconBlock"); err != nil {
+			return errors.Wrap(err, "broadcast and receive unseen sidecars")
+		}
+	}
+}
+
+func (s *Service) processDataColumnSidecarsFromExecutionFromColumnSidecar(ctx context.Context, sidecar blocks.VerifiedRODataColumn) error {
```

Comment:
this name seems too long, do we have other processDataColumnSidecars functions?

Reply:
Fixed in the latest version.